<a href="https://colab.research.google.com/github/adhikaryramen87/MachineLearning_Works/blob/main/Text_Vectorization_Demo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import all the required libraries

In [ ]:
import pandas as pd
import numpy as np

Load the data

In [ ]:
data = pd.read_csv('/content/spam.csv')
data.head()

,Category,Message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [ ]:
data.Category.value_counts()

,count
Category,
ham,4825
spam,747


In [ ]:
data['spam'] = data['Category'].apply(lambda x:1 if x == 'spam' else 0)
data.head()

,Category,Message,spam
0,ham,"Go until jurong point, crazy.. Available only ...",0
1,ham,Ok lar... Joking wif u oni...,0
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,1
3,ham,U dun say so early hor... U c already then say...,0
4,ham,"Nah I don't think he goes to usf, he lives aro...",0


Data Split

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data.Message, data.spam, stratify=data.spam, test_size=0.2, random_state=42)

In [ ]:
X_train[:4]

,Message
184,"He will, you guys close?"
2171,CAN I PLEASE COME UP NOW IMIN TOWN.DONTMATTER ...
5422,Ok k..sry i knw 2 siva..tats y i askd..
4113,"I'll see, but prolly yeah"


In [ ]:
X_train.shape

(4457,)

Message -> Feature Vectorization

In [ ]:
X_train.values

array(['He will, you guys close?',
       'CAN I PLEASE COME UP NOW IMIN TOWN.DONTMATTER IF URGOIN OUTL8R,JUST REALLYNEED 2DOCD.PLEASE DONTPLEASE DONTIGNORE MYCALLS,U NO THECD ISV.IMPORTANT TOME 4 2MORO',
       'Ok k..sry i knw 2 siva..tats y i askd..', ...,
       'Moby Pub Quiz.Win a £100 High Street prize if u know who the new Duchess of Cornwall will be? Txt her first name to 82277.unsub STOP £1.50 008704050406 SP Arrow',
       "Free entry in 2 a weekly comp for a chance to win an ipod. Txt POD to 80182 to get entry (std txt rate) T&C's apply 08452810073 for details 18+",
       'We are at grandmas. Oh dear, u still ill? I felt Shit this morning but i think i am just hungover! Another night then. We leave on sat.'],
      dtype=object)

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()
X_train_cv = vectorizer.fit_transform(X_train.values)
type(X_train_cv)

scipy.sparse._csr.csr_matrix

In [ ]:
X_train_cv

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 58884 stored elements and shape (4457, 7663)>

4457 => number of messages in the dataframe

7663 => number of unique words/features in the messages

In [ ]:
X_train_cv.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
vectorizer.get_feature_names_out()[3000:6000]

array(['fren', 'french', 'frens', ..., 'sha', 'shade', 'shadow'],
      dtype=object)

In [ ]:
vectorizer.vocabulary_

{'he': 3368,
 'will': 7434,
 'you': 7624,
 'guys': 3285,
 'close': 1856,
 'can': 1641,
 'please': 5205,
 'come': 1914,
 'up': 7102,
 'now': 4807,
 'imin': 3624,
 'town': 6922,
 'dontmatter': 2415,
 'if': 3603,
 'urgoin': 7128,
 'outl8r': 4976,
 'just': 3868,
 'reallyneed': 5553,
 '2docd': 387,
 'dontplease': 2416,
 'dontignore': 2414,
 'mycalls': 4640,
 'no': 4765,
 'thecd': 6743,
 'isv': 3754,
 'important': 3632,
 'tome': 6876,
 '2moro': 400,
 'ok': 4884,
 'sry': 6371,
 'knw': 3957,
 'siva': 6143,
 'tats': 6643,
 'askd': 1100,
 'll': 4152,
 'see': 5928,
 'but': 1587,
 'prolly': 5398,
 'yeah': 7596,
 'swing': 6590,
 'by': 1602,
 'in': 3643,
 'bit': 1375,
 'got': 3204,
 'some': 6245,
 'things': 6770,
 'to': 6854,
 'take': 6616,
 'care': 1662,
 'of': 4859,
 'here': 3415,
 'firsg': 2882,
 'shall': 6006,
 'book': 1435,
 'chez': 1783,
 'jules': 3859,
 'for': 2946,
 'half': 3305,
 'eight': 2554,
 'that': 6734,
 'with': 7465,
 'thanks': 6728,
 'your': 7628,
 'message': 4439,
 'really': 5552,


In [ ]:
vectorizer.get_feature_names_out()[1743]

'charges'

In [ ]:
y_train

,spam
184,0
2171,0
5422,0
4113,0
4588,0
...,...
1932,0
5316,0
2309,1
1904,1


In [ ]:
from sklearn.naive_bayes import MultinomialNB
model = MultinomialNB()
model.fit(X_train_cv, y_train)

MultinomialNB()

In [ ]:
X_test_cv = vectorizer.transform(X_test.values)
y_pred = model.predict(X_test_cv)
y_pred

array([0, 0, 0, ..., 0, 0, 0])

FP => Prediction = Positive Actual = Negative

Means the model predicts the mail is spam but actually the mail is important and urgent to respond.

In [ ]:
from sklearn.metrics import precision_score
precision = precision_score(y_test, y_pred)
print(precision)

0.9855072463768116


In [ ]:
from sklearn.pipeline import Pipeline
clf = Pipeline([
    ('vectorizer', CountVectorizer()),
    ('nb', MultinomialNB())
])

In [ ]:
clf.fit(X_train, y_train)

Pipeline(steps=[('vectorizer', CountVectorizer()), ('nb', MultinomialNB())])

In [ ]:
emails = [
    'Hi Priya, can we have a meeting tomorrow?',
    'Upto 20% OFF on H&M. Dont miss this exclusive reward'
]

emails_vectorizer = vectorizer.transform(emails)
model.predict(emails_vectorizer)

array([0, 1])

Raw TF

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

reviews = [
    'The plot is good.',
    'The plot is good, good, really good.'
]

tf_vectorizer = TfidfVectorizer(use_idf=False)

tf_matrix = tf_vectorizer.fit_transform(reviews)
tf_matrix.toarray()
df_tf_vectorizer = pd.DataFrame(tf_matrix.toarray(), columns = tf_vectorizer.get_feature_names_out())

df_tf_vectorizer

,good,is,plot,really,the
0,0.50000,0.50000,0.50000,0.00000,0.50000
1,0.83205,0.27735,0.27735,0.27735,0.27735


'is', 'the', 'a' [high frequency] => tf_vectorizer values will be also high

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

reviews = [
    'The plot is good.',
    'The plot is good, good, really good.'
    'The plot is a masterpiece',
    'I hate the plot',
    'The plot is okay'
]

tfidf_vectorizer = TfidfVectorizer()

tfidf_matrix = tfidf_vectorizer.fit_transform(reviews)
tfidf_matrix.toarray()
df_tfidf_vectorizer = pd.DataFrame(tfidf_matrix.toarray(), columns = tfidf_vectorizer.get_feature_names_out())

df_tfidf_vectorizer

,good,hate,is,masterpiece,okay,plot,really,the
0,0.628493,0.000000,0.508819,0.000000,0.000000,0.415993,0.000000,0.415993
1,0.700444,0.000000,0.378047,0.296142,0.000000,0.309078,0.296142,0.309078
2,0.000000,0.804612,0.000000,0.000000,0.000000,0.419880,0.000000,0.419880
3,0.000000,0.000000,0.456847,0.000000,0.715739,0.373502,0.000000,0.373502
